File: solutions_5658571765186560_0/C++/Lafsody/D-OminousOmino.cpp
#include<stdio.h>
#include<algorithm>

using namespace std;

int T,x,r,c,win;

int main(){
	int i,j;
	freopen("D-small.in","r",stdin);
    freopen("D-small.out","w",stdout);
//	freopen("D-large.in","r",stdin);
//    freopen("D-large.out","w",stdout);
	scanf("%d",&T);
	for(int t = 1; t <= T;t++){
		scanf("%d%d%d",&x,&r,&c);
		win = 0;
		int minrc = min(r,c);
		if((r * c) % x == 0 && (x + 1)/2 <= minrc){
			if(x <= 3){
				win = 1;
			}else if(x == 4){
				if(minrc != 2)win = 1;
			}
		}
		printf("Case #%d: %s\n",t,win==0?"RICHARD":"GABRIEL");
	}
}

File: solutions_5658571765186560_0/Python/Bhajun/D-ominous-omino.py
#!/usr/bin/env python

import sys


if __name__ == '__main__':
    T = int(sys.stdin.readline())
    
    for i in range(T):
        X, R, C = list(map(int, sys.stdin.readline().split()))
        R, C = min(R, C), max(R, C)

        # Richard can win if:
        #  - X is larger than 6 (in which case an omino with an inaccessible block)
        #  - The number of squares in the grid is not divisible by X
        #  - X is larger than R and C
        #  - X is more than twice as large as the smaller dimension
        if X > 6 or R * C % X > 0 or X > C or X > 2 * R:
            print('Case #%d: RICHARD' % (i + 1))

        # Gabriel can always win if X is less than 3
        elif X < 3:
            print('Case #%d: GABRIEL' % (i + 1))

        # Gabriel can win for higher X = 3, 4, 5, 6 with restrictions on the dimensions
        elif X == 3 and R >= 2 and (R % 3 == 0 or C % 3 == 0):
            print('Case #%d: GABRIEL' % (i + 1))

        elif X == 4 and R >= 3:
            print('Case #%d: GABRIEL' % (i + 1))

        # Otherwise, Richard wins
        else:
            print('Case #%d: RICHARD' % (i + 1))